Add resample and decimate utility methods for Analog and IrregularlysampledSignals

JuliaSprenger opened the pull request:
In the context of #526 it might be useful to add utility functions for adjusting the time sampling of `AnalogSignals` and `IrregularlySampledSignals`. This PR is wrapping the two scipy functions `decimate` and `resample` to provide a mechanism for decimating and resampling within Neo and adds some tests with it.

There is one issue coming with this PR: Scipy itself is not a Neo dependency yet, even though it is used in a couple of places for specific IOs. Here, I the new utility methods can only be used in case of scipy being available and will raise an ImportError if this is not the case.

Note: As scipy is not available in the Travis test setup, the new tests fail. Should we include scipy in the test environment / requirements for Neo or is this introducing too strong dependencies? Alternatively, these utility functions might be included in [elephant](https://github.com/neuralensemble/elephant), with the drawback ob being less visible to the users.

pep8speaks commented:
Hello @JuliaSprenger! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`neo/core/analogsignal.py`](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/core/analogsignal.py):

> [Line 552:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/core/analogsignal.py#L552): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 99 characters)
> [Line 558:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/core/analogsignal.py#L558): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (116 > 99 characters)
> [Line 575:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/core/analogsignal.py#L575): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 99 characters)
> [Line 587:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/core/analogsignal.py#L587): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 99 characters)
> [Line 588:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/core/analogsignal.py#L588): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (119 > 99 characters)
> [Line 593:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/core/analogsignal.py#L593): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (116 > 99 characters)
> [Line 612:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/core/analogsignal.py#L612): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (115 > 99 characters)

* In the file [`neo/core/irregularlysampledsignal.py`](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/core/irregularlysampledsignal.py):

> [Line 375:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/core/irregularlysampledsignal.py#L375): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 99 characters)
> [Line 376:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/core/irregularlysampledsignal.py#L376): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (119 > 99 characters)
> [Line 381:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/core/irregularlysampledsignal.py#L381): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (116 > 99 characters)
> [Line 400:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/core/irregularlysampledsignal.py#L400): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (117 > 99 characters)
> [Line 403:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/core/irregularlysampledsignal.py#L403): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (112 > 99 characters)
> [Line 404:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/core/irregularlysampledsignal.py#L404): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (113 > 99 characters)
> [Line 406:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/core/irregularlysampledsignal.py#L406): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (115 > 99 characters)

* In the file [`neo/test/coretest/test_analogsignal.py`](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_analogsignal.py):

> [Line 1163:38](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_analogsignal.py#L1163): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1173:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_analogsignal.py#L1173): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 99 characters)
> [Line 1174:64](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_analogsignal.py#L1174): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1180:38](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_analogsignal.py#L1180): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1191:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_analogsignal.py#L1191): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 99 characters)
> [Line 1192:48](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_analogsignal.py#L1192): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1192:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_analogsignal.py#L1192): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (109 > 99 characters)
> [Line 1198:38](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_analogsignal.py#L1198): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1203:30](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_analogsignal.py#L1203): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1204:52](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_analogsignal.py#L1204): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1205:67](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_analogsignal.py#L1205): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1210:44](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_analogsignal.py#L1210): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1210:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_analogsignal.py#L1210): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 99 characters)
> [Line 1216:38](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_analogsignal.py#L1216): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1222:54](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_analogsignal.py#L1222): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1223:58](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_analogsignal.py#L1223): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 1223:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_analogsignal.py#L1223): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (114 > 99 characters)
> [Line 1228:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_analogsignal.py#L1228): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 99 characters)
> [Line 1231:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_analogsignal.py#L1231): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 99 characters)

* In the file [`neo/test/coretest/test_irregularysampledsignal.py`](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_irregularysampledsignal.py):

> [Line 716:75](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_irregularysampledsignal.py#L716): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 717:100](https://github.com/NeuralEnsemble/python-neo/blob/630bb15fd6a5cbf7da5407908d174c8684f20514/neo/test/coretest/test_irregularysampledsignal.py#L717): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (109 > 99 characters)

JuliaSprenger commented:
@samuelgarcia @apdavison Do you have an opinion on the Scipy issue here?

apdavison commented:
For a long time I was against having scipy as a dependency, because it was often very slow and/or painful to install. Since the rise of  wheels and conda, this is no longer really a  problem, so I am ok with adding it as a dependency.

I think the name "decimate" is terrible, and "resample_regularly" could be just "resample". What do you think about having a single function "resample" for both AnalogSignal and IrrSampSig, which would use either scipy decimate or scipy resample under the hood, depending on the arguments?

samuelgarcia commented:
Hi julia.
This is a good idea.

I think we still use scipy as a weak dependency with the
```python
try:
   import scipy 
    HAVE_SCIPY= True
except importError
    HAVE_SCIPY= False
```
neo could be used in a web leigth weioght environement this could be usefull to not have it.
I don't known.
For desktop anybody have scipy.

apdavison commented on neo/core/irregularlysampledsignal.py:
```diff
@@ -353,7 +355,7 @@ def mean(self, interpolation=None):
             raise NotImplementedError
 
-    def resample(self, at=None, interpolation=None):
+    def resample_irregularly(self, at=None, interpolation=None):
```
What is the thinking behind providing a not-implemented method? Wouldn't it be better just to remove this?

JuliaSprenger replied:
I guess the idea was to make a suggestion for the future interface, so I kept it. But I am happy to remove it if you agree.

apdavison replied:
I think it would be better to remove it to avoid confusing users - but perhaps you could open an issue so as to remind us of the idea?

dizcza commented:
I was thinking to add resampling in Elephant, but I'm glad to see it's being implemented in Neo.
I agree with Andrew regarding (1) I also don't experience issues anymore with slow pip install \<whatever\> and (2) the same API - `resample` for both types of signal.
In the docs, it'd be nice to distinguish when to use `decimate` and when `resample`.

However, what I wanted to check and what I don't see here is some kind of tests that check for the artifacts of the resample/downsampled signal. It seems we, at our lab, are very cautious about adding spurious spikes as well as removing ones due to imperfect data processing. So I wonder, under what circumstances, if any, these methods can
* add spurious spikes;
* remove original spikes.

I was thinking of simulating raw analog signals from spikes, applying these methods, and thresholding back to see if the original spiketrain is preserved. Or testing on a chunk of real data.

JuliaSprenger commented:
@apdavison I removed the old resample interface and renamed the methods to `resample` and `downsample` and added more description and crossreferences in the docstring.
@samuelgarcia I updated the usage of scipy following your `HAVE_SCIPY` schema.
@dizcza I added two sentences in the docstrings making the difference between the two functions more clear. However, I think that the tests you are suggesting should be another PR, since the problems you are describing highly depend on the type of data (sources) you are looking at and will probably lead to longer discussion, especially since the process of spike sorting you are hinting at is not part of Neo.
